New hypercalls

dsemenets opened the pull request:
Added hypercalls:
 - xen_version
 - sysctl
Added getvcpuinfo to domctl hypercall

oleksiimoisieiev commented on include/zephyr/xen/public/physdev.h:
```diff
@@ -0,0 +1,383 @@
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Copyright (c) 2006, Keir Fraser
+ */
```
Please leave a comment about where this source code was taken from

lorc commented:
Could you please provide a good commit message? I have shared guidelines multiple times already.

Also, I want to note that you actually adding only two new hypercalls: `xen_version` and `sysctl`.  All other points in your commit message are either specific domctl or sysctl. It is crucial to use correct terminology.

xakep-amatop commented:
Could you please fix all errors detected during compliance checks (except new typedefs)?
https://github.com/xen-troops/zephyr/actions/runs/4671475965/jobs/8272516208?pr=22

xakep-amatop commented:
according to daily-sync status of @firscity 
almost all xen public headers will be moved to xenlib
so, I have a question, do we really need this changes in Zephyr repo?

dsemenets commented:
> I have a question, do we really need this changes in Zephyr repo?

I guess this change could be moved to the xenlib

dsemenets commented:
> Could you please fix all errors detected during compliance checks (except new typedefs)?

Should I fix everything? Including errors (for example leading space) introduced previously?

xakep-amatop commented:
> > Could you please fix all errors detected during compliance checks (except new typedefs)?
> 
> Should I fix everything? Including errors (for example leading space) introduced previously?

it will be enough to fix only errors produced by yours changes
errors introduced previously are already fixed here: https://github.com/xen-troops/zephyr/pull/35

dsemenets commented:
> it will be enough to fix only errors produced by yours changes

It will be very unreadable code... 
:(

firscity commented on include/zephyr/xen/public/sysctl.h:
```diff
@@ -0,0 +1,174 @@
+#include <zephyr/sys/util_macro.h>
+#include "xen.h"
+#include "domctl.h"
+
```
Please add the same as Xen protection to include it only for Dom0-related code

> #if !defined(__XEN__) && !defined(__XEN_TOOLS__)
#error "sysctl operations are intended for use by node control tools only"
#endif

As a condition you can use `CONFIG_XEN_INITIAL_DOMAIN`. You can use `domctl.h` as example.

firscity commented on include/zephyr/xen/public/sysctl.h:
```diff
@@ -0,0 +1,174 @@
+/******************************************************************************
```
Please add same SPDX identifier for line 1 as other public headers (MIT).

MAJOR: please take this header from stable XEN release (4.17.0), it looks different and already has this change.

firscity commented on include/zephyr/xen/public/version.h:
```diff
@@ -0,0 +1,117 @@
+/******************************************************************************
```
Please add same SPDX identifier for line 1 as other public headers (MIT).

MAJOR: please take this header from stable XEN release (4.17.0), it looks different and already has this change.

firscity commented on include/zephyr/xen/public/sysctl.h:
```diff
@@ -0,0 +1,174 @@
+
+#endif /* __XEN_PUBLIC_SYSCTL_H__ */
+
+/*
```
Please, remove this lines, they are redundant for Zephyr.

firscity commented on include/zephyr/xen/public/version.h:
```diff
@@ -0,0 +1,117 @@
+
+#endif /* __XEN_PUBLIC_VERSION_H__ */
+
+/*
```
Please, remove this lines, they are redundant for Zephyr.

firscity commented on arch/arm64/core/xen/hypercall.S:
```diff
@@ -25,3 +25,5 @@ HYPERCALL(event_channel_op);
 HYPERCALL(memory_op);
 HYPERCALL(domctl);
+HYPERCALL(xen_version);
+HYPERCALL(sysctl);
```
Please, cover this with `CONFIG_XEN_INITIAL_DOMAIN`, we missed the same for domctl, it will be added in `zephyr-v3.3.0-xt` branch.

dsemenets replied:
should I rebase this to v3.3.0-xt?

firscity commented on include/zephyr/arch/arm64/hypercall.h:
```diff
@@ -15,5 +15,7 @@ int HYPERVISOR_hvm_op(int op, void *param);
 int HYPERVISOR_grant_table_op(int op, void *uop, unsigned int count);
 int HYPERVISOR_domctl(void *param);
+int HYPERVISOR_xen_version(int op, void *param);
+int HYPERVISOR_sysctl(void *param);
```
Please, cover this with `CONFIG_XEN_INITIAL_DOMAIN`, we missed the same for domctl, it will be added in `zephyr-v3.3.0-xt` branch.

firscity reviewed (commented):
Also, please, add more about motivation to commit message (`These call were absent... They are used for xstat functionality... gather domain data...` etc.)

lorc commented on arch/arm64/core/xen/hypercall.S:
```diff
@@ -23,5 +23,9 @@ HYPERCALL(grant_table_op);
 HYPERCALL(event_channel_op);
 HYPERCALL(hvm_op);
-HYPERCALL(memory_op);
+HYPERCALL(memory_op)
```
Why did you removed semicolon?

dsemenets replied:
Missprint

lorc commented on drivers/xen/dom0/domctl.c:
```diff
@@ -313,6 +313,22 @@ int xen_domctl_cacheflush(int domid,  struct xen_domctl_cacheflush *cacheflush)
 	return do_domctl(&domctl);
 }
 
+int xen_domctl_getvcpu(int domid, int vcpu, struct xen_domctl_getvcpuinfo *info)
```
why are you using `int` types for `domid` and `vcpu`? They can't be negative.

Okay, I see that surrounding code uses `int` for `domid`, so you can leave it for consistency (but we will need to fix this in the future). But at least you can change type for `vcpu` to something more appropriate.

dsemenets replied:
Done

lorc commented on drivers/xen/dom0/sysctl.c:
```diff
@@ -0,0 +1,46 @@
+	return ret;
+}
+
+int xen_sysctl_getdomaininfo(struct xen_domctl_getdomaininfo *domaininfo, int first, int num)
```
`int` suggests that `first` and `num` can be negative. This is not true.

dsemenets replied:
Done

lorc reviewed (changes requested):
Please rewrite both commit messages according to the requirements.

klogg commented:
@dsemenets what's left here? can this be pushed or some changes are still pending?

dsemenets commented:
@klogg Nothing from my side. Sorry I forget to push latest changes

firscity commented on scripts/checkpatch/typedefsfile:
```diff
@@ -6,3 +6,5 @@ pinctrl_soc_pin_t
 io_rw_32
 \b[a-zA-Z_][a-zA-Z0-9_]*TypeDef
 Pwm
+xen_sysctl_cpuinfo_t
```
This patch is redundant. From my experience - patches with new typedef is OK for upstream.
Comments about `checkpatch issues` were related to mixed tabs/spaces issues, that were present in previous versions.

firscity reviewed (commented):
Commit message for "xen: New hypercalls added for xstat library" still not "upstream quality". xstat is not a part of Zephyr kernel, you should add more description about this maybe with links. Please, consider that all of this patches should be upstreamed to Zephyr repository and maintainer/reviewer should clearly understand what is this from commit message.

dsemenets commented:
Removed patch for file scripts/checkpatch/typedefsfile
Fixed commit message

lorc commented:
It's a pity that we can't leave comments to commit messages (like in the mailing list), so I'll do this there:

> xen: Add new hypercalls

Too generic. This can include any hypercall.

> Added hypercalls:
> - xen_version
> - sysctl getphysinfo
> - sysctl getdomaininfolist
> - sysctl getcpuinfo
>Added getvcpuinfo to domctl hypercall

Missing full stop.

> Those calls are used for gathering information about running domains

Missing full stop.

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>


How about this?


```
xen: add statistics-related hypercalls

Add couple of Xen hypercalls that can be used get and present
information about Xen hypervisor, physical machine and running guests:

- "xen_version" - to get Xen version.

- sysctl hypercall "getphysinfo" allows read information about
  physical machine: number CPUs, memory sizes, hardware capabilities,
  etc.

- sysctl hypercall "xen_sysctl_getdomaininfolist" returns array of
  domain info structures that provide information about particular
  domain(s).

- sysctl hypercall "get_cpu_info" provides information about
  physical CPU(s).

- domctl hypercall "getvcpu" provides information about domain's
  virtual CPU(s).

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
```


Now, looking at this proper commit message, I'd suggest to split this into three separate commits: 

 - Add xen_version hypercall
 - Add domctl getvcpu hypercall
 - Add sysctl machinery + required call 

Because right now this one commit has multiple changes that are hard to call atomic. Also, please note that it is much easier to review three separate self-contained commits.

dsemenets commented:
I have divided this PR on 3 separate commits

xakep-amatop commented on include/zephyr/xen/public/version.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef __XEN_PUBLIC_VERSION_H__
+#define __XEN_PUBLIC_VERSION_H__
+
+#if !defined(CONFIG_XEN_INITIAL_DOMAIN)
```
(note) you can use `#ifndef` instead `#if !defined`

dsemenets replied:
I prefer to use `#if !defined`.

xakep-amatop replied:
then why have you used ifndef a few lines above in this file?

dsemenets replied:
1. `#ifndef` and `#define` have same length
2. That is template

xakep-amatop replied:
ok, in any case for me it is only some kind of cosmetic changes
so, I'm not against to leave it as is

xakep-amatop commented on drivers/xen/dom0/domctl.c:
```diff
@@ -313,6 +313,22 @@ int xen_domctl_cacheflush(int domid,  struct xen_domctl_cacheflush *cacheflush)
 
+int xen_domctl_getvcpu(int domid, uint32_t vcpu, struct xen_domctl_getvcpuinfo *info)
+{
+	xen_domctl_t domctl;
```
(note) you can rewrite it in the next way and remove memset:
```
xen_domctl_t domctl = {
        .cmd = XEN_DOMCTL_getvcpuinfo,
	.domain = domid,
	.u.getvcpuinfo.vcpu = vcpu,
};
```

dsemenets replied:
Ok

xakep-amatop commented:
@dsemenets pls, fix errors detected by compliance check:
```
 Commit 5d0f4e174d:
12: UC4 Commit body line exceeds max length (77>75): "- sysctl hypercall "get_cpu_info" provides information about physical CPU(s)."
```

xakep-amatop commented on drivers/xen/dom0/sysctl.c:
```diff
@@ -0,0 +1,47 @@
+
+int xen_sysctl_physinfo(struct xen_sysctl_physinfo *info)
+{
+	xen_sysctl_t sysctl = {0};
```
(cosmetic)
```
xen_sysctl_t sysctl = {
    .cmd = XEN_SYSCTL_physinfo,
};
```

dsemenets replied:
Done

xakep-amatop commented on drivers/xen/dom0/sysctl.c:
```diff
@@ -0,0 +1,47 @@
+int xen_sysctl_getdomaininfo(struct xen_domctl_getdomaininfo *domaininfo,
+			     uint16_t first, uint16_t num)
+{
+	xen_sysctl_t sysctl = {0};
```
(cosmetic) the same as above

dsemenets replied:
Done

xakep-amatop commented:
(cosmetic) pls, fix typo in the second commit message:
```
This hypercall can be used get some information about physical machine and
running duests:
```
Use `guests` instead of `duests`

dsemenets commented:
Fixes in systcl.c and domctl.c
Fixed misprint in commit message

lorc commented on include/zephyr/xen/public/version.h:
```diff
@@ -0,0 +1,31 @@
+#define __XEN_PUBLIC_VERSION_H__
+
+#if !defined(CONFIG_XEN_INITIAL_DOMAIN)
+#error "sysctl operations are intended for use by node control tools only"
```
Why is this? `XENVER_` calls are available for all guests.

dsemenets replied:
Done

lorc commented on include/zephyr/arch/arm64/hypercall.h:
```diff
@@ -14,6 +14,9 @@ int HYPERVISOR_event_channel_op(int op, void *param);
 int HYPERVISOR_hvm_op(int op, void *param);
 int HYPERVISOR_memory_op(int op, void *param);
 int HYPERVISOR_grant_table_op(int op, void *uop, unsigned int count);
+#ifdef CONFIG_XEN_INITIAL_DOMAIN
```
This is not needed. Xen has own access system. Moreover, it is possible to configure XSM to allow other guests to issue `domctl` commands.

`XENVER_` calls are available or all domains by default.

dsemenets replied:
Done

firscity replied:
@lorc agree with xen_version, but domctl and sysctl are unclear. I'm not sure about hypercall, but all wrapper functions for sysctl and domctl have a build check if they are built in control domain.

dsemenets replied:
@lorc @firscity Should I leave this or remove?

firscity replied:
I think we should leave it as libxl headers say, only for control domain.
Please move `int HYPERVISOR_sysctl(void *param);` under `#ifdef`.

dsemenets commented:
Fixed issues from @lorc @xakep-amatop, corrected commit message

xakep-amatop reviewed (approved):
```
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
```

lorc reviewed (approved):
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

dsemenets commented:
Updated commit messages with `Acked-by:`

dsemenets committed:
xen: Add xen version hypercall

This hypercall used to obtain version of xen hypervisor

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>

dsemenets committed:
xen: Add sysctl hypercall

This hypercall can be used get some information about physical machine
and running guests:

- sysctl hypercall "xen_sysctl_getphysinfo" allows read information about
physical machine: number CPUs, memory sizes, hardware capabilities, etc.

- sysctl hypercall "xen_sysctl_getdomaininfolist" returns array of domain
info structures that provide information about particular domain(s).

- sysctl hypercall "get_cpu_info" provides information about physical
CPU(s).

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>

dsemenets committed:
xen: Added getvcpuinfo function in domctl hypercall

domctl hypercall "getvcpu" provides information about domain's
virtual CPU(s).

Signed-off-by: Dmytro Semenets <dmytro_semenets@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>

dsemenets commented:
Rebased on zephyr-v3.3.0-xt branch

firscity reviewed (commented):
With last comment: `Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
